### PR TITLE
Run Mono's System.Drawing on .NET Core

### PR DESCRIPTION
**Pull request by qmfrederik**
Except for Point, Size and Rectangle, .NET Core doesn't have drawing primitives (such as color - dotnet/corefix#2315) nor drawing capabilities.

@migueldeicaza suggested (dotnet/corefx#2020) to use Mono's System.Drawing assembly and run it on .NET Core, and @akoeplinger partially implemented this, though the integration became outdated.

I've done more work on this and have ported Mono's System.Drawing to .NET Core rc2 (using some of the rc3 packages) - see https://github.com/qmfrederik/CoreCompat/.

It required some changes in the Mono implementation and I'm submitting this PR to have them accepted into the Mono repository.

The changes can be summarized as:
- Some `#if`s to workaround functionality which is not available in .NET Core (say, Serialization)
- Updates to work with the `string`, reflection, `Stream` and `MemoryStream` API changes (most of them backwards-compatible with existing .NET Code)
- Removal of the CAS attributes on the unit tests for .NET 4 and above
- When marshaling, change `Charset.Auto` to `CharSet.Unicode` and use `UnmanagedType.Interface` for COM types
- Update the code that detects on which platform we're running

I've tried to keep the commits very granular to make them easy to review - happy to squash or rewrite them if needed.

All feedback welcome,

### CODE REVIEW

**Comment by migueldeicaza** [comment]
It is unlikely that we will take the marshall changes, it would be a breaking change.

Most of the changes are related to .NET Core having a limited API, which is being addressed.

File: mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/PathGradientBrushTest.cs [hunk 1]
```diff
@@ -36,7 +36,9 @@
 namespace MonoTests.System.Drawing.Drawing2D {
 
 	[TestFixture]
+#if !NET40
 	[SecurityPermission (SecurityAction.Deny, UnmanagedCode = true)]
+#endif
```
Comment (akoeplinger) [inline mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/PathGradientBrushTest.cs]: We're using the `NET_4_0` symbol in the codebase, but given that CAS was never fully implemented on Mono the easier/cleaner solution is just to remove the attributes.

**Comment by qmfrederik** [comment]
@migueldeicaza Happy to split the PR in changes you can take and changes you cannot take.

Regarding the Marshall changes, do you refer to the Auto vs. Unicode change or the custom marshaler? CharSet.Auto is not available on .NET Core, so what would be the alternative, an `#if`?

Regarding the marshaler,
https://github.com/mono/mono/blob/ef407901f8fdd9ed8c377dbec8123b5afb932ebb/mcs/class/System.Drawing/Systemy.Drawing/ComIStreamMarshaler.cs#L47 states: "This custom marshaler should be replaced with UnmanagedType.Interface".

File: mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/PathGradientBrushTest.cs (hunk 1 above)
Comment (qmfrederik) [inline mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/PathGradientBrushTest.cs]: OK, I just pushed a new commit which removes the CAS attributes in the unit tests.

Let me know if it's easier I prepare a seperate PR for that.

**Comment by migueldeicaza** [comment]
.NET Core needs to be fixed to bring back .Auto.   This is not going to cause trouble here, it is going to cause trouble everywhere.

**Comment by migueldeicaza** [comment]
FWIW - they also need to bring UTF8 encoding as a marshaling value, Auto/Ansi are band-aids that were there because UTF8 was never added.   The MS team has agreed to add the functionality, but it has been postponed from the current sprint, as it touches on too many other things.

**Comment by qmfrederik** [comment]
Closing in favor of #4483 which targets netcoreapp2.0, the port is much cleaner now.

**Pull request closed without merging**